TOOLS/lua/autocrop.lua

```diff
@@ -18,97 +18,225 @@
-        )
-        mp.msg.error(
-            "Does your version of ffmpeg/libav support AVFrame metadata?"
+        mp.msg.error("No crop data.")
+        mp.msg.info("Was the cropdetect filter successfully inserted?")
+        mp.msg.info("Does your version of ffmpeg/libav support AVFrame metadata?")
+        return
+    end
+
+
+    -- Verify that the metadata meets the requirements and convert it.
+    if meta.w and meta.h and meta.x and meta.y then
+        meta = {
+            w = tonumber(meta.w),
+            h = tonumber(meta.h),
+            x = tonumber(meta.x),
+            y = tonumber(meta.y),
+            min_w = mp.get_property_native("width") * options.detect_min_ratio,
+            min_h = mp.get_property_native("height") * options.detect_min_ratio,
+            max_w = mp.get_property_native("width"),
+            max_h = mp.get_property_native("height")
+        }
+    else
+        mp.msg.error("Got empty crop data.")
+        mp.msg.info("You might need to increase detect_seconds.")
+        return
+    end
+
+
+    apply_crop(meta)
+end
+
+
+function apply_crop(meta)
+
+    -- Verify if it is necessary to crop.
+    local is_effective = meta.x > 0 or meta.y > 0
+        or meta.w < meta.max_w or meta.h < meta.max_h
+
+    if not is_effective then
+        mp.msg.info("No area detected for cropping.")
+        return
+    end
+
+
+    -- Verify if it is over cropped.
+    local is_not_excessive = meta.w >= meta.min_w and meta.h >= meta.min_h
+
+    if not is_not_excessive then
+        mp.msg.info("The area to be cropped is too large.")
+        mp.msg.info("You might need to decrease detect_min_ratio.")
+        return
+    end
+
+
+    -- Remove existing crop.
+    remove_filter(label.crop)
+
+
+    -- Apply crop.
+    mp.command(
+        string.format("vf pre @%s:lavfi-crop=w=%s:h=%s:x=%s:y=%s",
+            label.crop, meta.w, meta.h, meta.x, meta.y
         )
+    )
+end
+
+
+function on_start()
+
+    -- Remove existing cropdetect.
+    remove_filter(label.cropdetect)
+
+
+    -- If auto is not true, exit.
+    if not options.auto then
+        return
+    end
+
+
+    -- Verify if there is a crop, decide whether to keep it by auto_keep.
+    if is_filter_exist(label.crop)
+        and options.auto_keep
+    then
+        mp.msg.info("Keep the existing crop.")
+        return
+    end
+
+
+    -- Verify if there is enough time for autocrop.
+    local time_spent = options.detect_seconds
+    local is_time_enough = time_spent < mp.get_property_native("playtime-remaining")
+
+    if not is_time_enough then
+        mp.msg.warn("Not enough time to crop.")
     end
-    -- remove the cropdetect filter
-    del_filter_if_present(cropdetect_label)
-    timer=nil
+
+
+    -- Do delectcrop.
+    detect_crop()
 end
 
-mp.add_key_binding("C", "auto_crop", autocrop_start)
+
+function on_toggle()
+
+    -- Cropped => Remove it.
+    if remove_filter(label.crop) then
+        return
+    end
+
+
+    -- Detecting => Wait it.
```

Comment:
"wait it" -> "leave it"

the code isn't actually waiting in this case, but aborting.